### PR TITLE
[feature](nereids) push down Project through Limit

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: noissue

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/RewriteJob.java [hunk 1]
```diff
@@ -64,6 +66,7 @@ public RewriteJob(CascadesContext cascadesContext) {
                 .add(topDownBatch(ImmutableList.of(new NormalizeAggregate())))
                 .add(topDownBatch(ImmutableList.of(new ColumnPruning())))
                 .add(topDownBatch(ImmutableList.of(new AggregateDisassemble())))
+                .add(topDownBatch(ImmutableList.of(new SwapLimitAndProject())))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/RewriteJob.java:69]: nit: rename to PushDownProjectThroughLimit

**Comment by wangshuo128** [comment]
Just curious, does BE support limit output before calculating projections inside a BE plan node currently?
Let's say we have a plan `limit -> project -> olap scan`. After applying the rule, it's rewritten to `project -> limit -> olap scan`.  
The translated plan would be `OlapScanNode(project and then limit)` or something else. If so, does BE support this plan(first limit and then project inside olap scan) currently? @Kikyou1997 @morrySnow

**Review by morrySnow — APPROVED** [review]

**Comment by morrySnow** [comment]
> Just curious, does BE support limit output before calculating projections inside a BE plan node currently? Let's say we have a plan `limit -> project -> olap scan`. After applying the rule, it's rewritten to `project -> limit -> olap scan`. The translated plan would be `OlapScanNode(scan data from underlying olap scanners, limit and then project)` or something else. If so, does BE support this plan(firstly limit and then project inside olap scan) currently? @Kikyou1997 @morrySnow

This rule is rewrite `project -> limit` to `limit -> project`. The reason is we could get tree like `project -> limit -> project -> other node`. If we do not rewrite it. we could not merge the two project into one. And if we has more than one project on one node, the second one will overwrite the first one when translate. Then, be will core dump or return `slot cannot find` error.

**Comment by wangshuo128** [comment]
> > Just curious, does BE support limit output before calculating projections inside a BE plan node currently? Let's say we have a plan `limit -> project -> olap scan`. After applying the rule, it's rewritten to `project -> limit -> olap scan`. The translated plan would be `OlapScanNode(scan data from underlying olap scanners, limit and then project)` or something else. If so, does BE support this plan(firstly limit and then project inside olap scan) currently? @Kikyou1997 @morrySnow
> 
> This rule is rewrite `project -> limit` to `limit -> project`. The reason is we could get tree like `project -> limit -> project -> other node`. If we do not rewrite it. we could not merge the two project into one. And if we has more than one project on one node, the second one will overwrite the first one when translate. Then, be will core dump or return `slot cannot find` error.

ok, thanks. :)

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/RewriteJob.java [hunk 2]
```diff
@@ -64,7 +66,8 @@ public RewriteJob(CascadesContext cascadesContext) {
                 .add(topDownBatch(ImmutableList.of(new NormalizeAggregate())))
                 .add(topDownBatch(ImmutableList.of(new ColumnPruning())))
                 .add(topDownBatch(ImmutableList.of(new AggregateDisassemble())))
-                .add(topDownBatch(ImmutableList.of(new SwapFilterAndProject())))
+                .add(topDownBatch(ImmutableList.of(new PushdownLimitThroughProject())))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/RewriteJob.java:69]:
```suggestion
                .add(topDownBatch(ImmutableList.of(new PushdownProjectThroughLimit())))
```

**Review by morrySnow — CHANGES_REQUESTED** [review]

**Comment by wangshuo128** [comment]
BTW, we'd better correct the PR title as well.

**Comment by morrySnow** [comment]
> BTW, we'd better correct the PR title as well.

done😁

**Review by morrySnow — APPROVED** [review]

**Pull request merged**